sdk/core/System.ClientModel/src/Message/BinaryContent.cs

```diff
@@ -191,4 +200,45 @@ public override void Dispose()
             }
         }
     }
+
+    private sealed class StreamBinaryContent : BinaryContent
+    {
+        private readonly Stream _stream;
+        private readonly long _origin;
+
+        public StreamBinaryContent(Stream stream)
+        {
+            if (!stream.CanSeek)
+            {
+                throw new ArgumentException("Stream must be seekable.", nameof(stream));
+            }
+
+            _stream = stream;
+            _origin = stream.Position;
+        }
+
+        public override bool TryComputeLength(out long length)
+        {
+            // CanSeek is validated in constructor - it will always be true.
+            length = _stream.Length - _origin;
+            return true;
+        }
+
+        public override void WriteTo(Stream stream, CancellationToken cancellationToken)
+        {
+            _stream.Seek(_origin, SeekOrigin.Begin);
+            _stream.CopyTo(stream, cancellationToken);
```

Comment:
Note - using our own polyfill here has us throwing `OperationCanceledException` where Azure.Core throws `TaskCanceledException`.  This is recommended for .NET BCL types (i.e. it is the guidance from the Framework Design Guidelines).  We will want to consider whether it is acceptable to make this change in Azure.Core as well.

Reply:
Tagging as a consideration in:
- https://github.com/Azure/azure-sdk-for-net/pull/42386
- https://github.com/Azure/azure-sdk-for-net/issues/42757